pkg/agent/client/client.go

```diff
@@ -197,30 +199,34 @@ func (c *client) FetchJWTSVID(ctx context.Context, jsr *node.JSR) (*JWTSVID, err
 	}, nil
 }
 
+// Release the underlying connection.
 func (c *client) Release() {
-	c.m.Lock()
-	defer c.m.Unlock()
-
-	if c.conn != nil {
-		c.conn.Close()
-		c.conn = nil
-	}
+	c.nodeConn.MarkForRelease()
 }
 
-func (c *client) newNodeClient(ctx context.Context) (node.NodeClient, error) {
-	if c.newNodeClientCallback != nil {
-		return c.newNodeClientCallback()
-	}
+type releaser interface {
+	Release()
+	MarkForRelease()
+}
 
+func (c *client) newNodeClient(ctx context.Context) (node.NodeClient, releaser, error) {
 	c.m.Lock()
 	defer c.m.Unlock()
 
-	if c.conn == nil {
+	// close a previous connection that was marked for release
+	if c.nodeConn != nil && c.nodeConn.IsMarkedForRelease() {
+		c.nodeConn = nil
+		c.nodeConn.Release()
```

Comment:
this line will panic since you've just set `c.nodeConn` to `nil`

Reply:
hmm, will fix and add some coverage for this